modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/DenseVectorFieldMapper.java

```diff
@@ -0,0 +1,193 @@
+import org.elasticsearch.common.xcontent.XContentParser.Token;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.query.QueryShardContext;
+import org.elasticsearch.search.DocValueFormat;
+import org.joda.time.DateTimeZone;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+
+/**
+ * A {@link FieldMapper} for indexing a dense vector of floats.
+ */
+
+public class DenseVectorFieldMapper extends FieldMapper implements ArrayValueMapperParser {
+
+    public static final String CONTENT_TYPE = "dense_vector";
+    public static short MAX_DIMS_COUNT = 500; //maximum allowed number of dimensions
+    private static final byte INT_BYTES = 4;
+
+    public static class Defaults {
+        public static final MappedFieldType FIELD_TYPE = new DenseVectorFieldType();
+
+        static {
+            FIELD_TYPE.setTokenized(false);
+            FIELD_TYPE.setIndexOptions(IndexOptions.NONE);
+            FIELD_TYPE.setHasDocValues(true);
+            FIELD_TYPE.setOmitNorms(true);
+            FIELD_TYPE.freeze();
+        }
+    }
+
+    public static class Builder extends FieldMapper.Builder<Builder, DenseVectorFieldMapper> {
+
+        public Builder(String name) {
+            super(name, Defaults.FIELD_TYPE, Defaults.FIELD_TYPE);
+            builder = this;
+        }
+
+        @Override
+        public DenseVectorFieldType fieldType() {
+            return (DenseVectorFieldType) super.fieldType();
+        }
+
+        @Override
+        public DenseVectorFieldMapper build(BuilderContext context) {
+            setupFieldType(context);
+            return new DenseVectorFieldMapper(
+                    name, fieldType, defaultFieldType,
+                    context.indexSettings(), multiFieldsBuilder.build(this, context), copyTo);
+        }
+    }
+
+    public static class TypeParser implements Mapper.TypeParser {
+        @Override
+        public Mapper.Builder<?,?> parse(String name, Map<String, Object> node, ParserContext parserContext) throws MapperParsingException {
+            DenseVectorFieldMapper.Builder builder = new DenseVectorFieldMapper.Builder(name);
+            return builder;
+        }
+    }
+
+    public static final class DenseVectorFieldType extends MappedFieldType {
+
+        public DenseVectorFieldType() {}
+
+        protected DenseVectorFieldType(DenseVectorFieldType ref) {
+            super(ref);
+        }
+
+        public DenseVectorFieldType clone() {
+            return new DenseVectorFieldType(this);
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+
+        @Override
+        public DocValueFormat docValueFormat(String format, DateTimeZone timeZone) {
+            throw new UnsupportedOperationException("[dense_vector] field doesn't support doc values");
```

Comment:
A better message might be to say that it doesn't support `docvalue_fields` and aggregations (which are the two features that use this API. Let's also include the field name in the message?